Fix avg_pool3d precision for BFloat16/Half

vishalgoyal316 opened the pull request:
Use higher precision accumulator (opmath_type) for sum computation

in adaptive_avg_pool3d CPU implementation to fix numerical errors and CPU-GPU inconsistencies with BFloat16/Half dtypes.

- Accumulate in float precision instead of BFloat16/Half
- Compute divide_factor as single multiplication vs sequential divisions
- Add test for CPU-GPU parity with reduced precision

Fixes #140696

vishalgoyal316 committed:
Fix adaptive_avg_pool3d precision for BFloat16/Half Use higher precision accumulator (opmath_type) for sum computation
in adaptive_avg_pool3d CPU implementation to fix numerical errors
and CPU-GPU inconsistencies with BFloat16/Half dtypes.

- Accumulate in float precision instead of BFloat16/Half
- Compute divide_factor as single multiplication vs sequential divisions
- Add test for CPU-GPU parity with reduced precision

vishalgoyal316 commented:
@pytorchbot label "topic: not user facing"

Skylion007 commented on test/nn/test_pooling.py:
```diff
@@ -1011,6 +1011,49 @@ def test_adaptive_avg_pool3d_output_size_one(self, device):
         c = out.size(1)
         self.assertEqual(out.stride(), [c, 1, 1, 1, 1])
 
+    @onlyCUDA
```
Hmm, why only cuda, why not just skip CPU? Good to test RoCM, MPS, and XPU too, right?

vishalgoyal316 replied:
Thanks for your comments. You are right, I have added test support for all GPUs.
Please review and let me know if you have any other suggestions.

vishalgoyal316 committed:
Updated testcase test_adaptive_avg_pool3d_cpu_gpu_parity to support CUDA, MPS, XPU and HPU

Skylion007 commented on test/nn/test_pooling.py:
```diff
@@ -1011,6 +1011,57 @@ def test_adaptive_avg_pool3d_output_size_one(self, device):
+        "Requires at least one accelerator device",
+    )
+    @dtypes(torch.bfloat16, torch.half)
+    def test_adaptive_avg_pool3d_cpu_gpu_parity(self, device, dtype):
```
How come the nn.functional.adaptive_avg_pool3d test on `torch/testing/_internal/common_methods_invocations.py` doesn't catch this? Is the tolerance too high?

vishalgoyal316 replied:
The tolerance contributes to the issue, but the primary reason is that randomly generated inputs don't reliably expose this particular precision accumulation edge case.

Skylion007 commented on test/nn/test_pooling.py:
```diff
@@ -1011,6 +1011,57 @@ def test_adaptive_avg_pool3d_output_size_one(self, device):
+        result_cpu = F.adaptive_avg_pool3d(input_tensor.cpu(), output_size)
+        result_accelerator = F.adaptive_avg_pool3d(input_tensor.to(device), output_size)
+
+        self.assertEqual(result_cpu, result_accelerator.cpu(), atol=1e-2, rtol=1e-3)
```
should this call assert close?

vishalgoyal316 committed:
Convert adaptive_avg_pool3d parity test to CPU-only precision test

Compare bfloat16/half results against float32 reference on CPU instead of requiring GPU for cross-device validation

Skylion007 commented:
Does this test fail on CUDA? Why not just enable it everywhere though to test other kernels too

Skylion007 commented on test/nn/test_pooling.py:
```diff
@@ -1011,6 +1011,53 @@ def test_adaptive_avg_pool3d_output_size_one(self, device):
 
+    @onlyCPU
+    @dtypes(torch.bfloat16, torch.half)
+    def test_adaptive_avg_pool3d_reduced_precision(self, device, dtype):
```
Why not test this on all devices which support these dtypes?

Skylion007 commented:
Ah found it: https://github.com/pytorch/pytorch/pull/136116

vishalgoyal316 committed:
Fix bfloat16/half precision for adaptive_avg_pool3d and avg_pool3d

Use at::opmath_type for higher precision accumulation in both adaptive_avg_pool3d
and avg_pool3d CPU implementations. Enables bfloat16/half support in legacy avg_pool3d.
Fixes #140696.

vishalgoyal316 commented:
> This is also partially my fault lol: #136091

I have taken care this failure in this PR. Thanks!

vishalgoyal316 committed:
Add float16/bfloat16 to avg_pool3d OpInfo dtypes for CPU

vishalgoyal316 committed:
Skip test_adaptive_avg_pool3d_reduced_precision on MPS

MPS doesn't have native adaptive_avg_pool3d implementation.
The test expects the CPU precision fix which doesn't apply to MPS fallback.

Related to #140696

Skylion007 commented:
Does the MPS kernel have the same bug?

vishalgoyal316 committed:
Fix MPS avg_pool3d backward precision for BFloat16/Half

The MPS backward pass was performing division in low precision (T) instead
of high precision (opmath_t), causing numerical errors and test failures
for half/bfloat16 dtypes.

Changes:
- Use opmath_t<T> for divisor and gradient division in backward pass
- Remove @expectedFailureMPS since the bug is now fixed
- Makes MPS backward consistent with forward pass and CPU implementation

Fixes test_output_grad_match_nn_functional_avg_pool3d_mps_float16 failure.

Related to #140696

vishalgoyal316 commented:
> Does the MPS kernel have the same bug?

@Skylion007, you are right, I have fixed it now.

vishalgoyal316 commented:
## Updated with MPS precision fix

The MPS test failures were due to a real precision bug in the backward pass. 
I've fixed it by using `opmath_t<T>` for division (matching CPU and forward pass).

**Changes:**
- `aten/src/ATen/native/mps/kernels/Pooling.metal`: Use high precision for backward division
- `test/nn/test_pooling.py`: Removed `@expectedFailureMPS` - test should now pass

**Note:** The `test_pynode_destruction_deadlock` failure is unrelated to this PR 
(autograd test, not pooling). This is a known flaky test.

vishalgoyal316 committed:
Skip test_adaptive_avg_pool3d_reduced_precision on MPS

MPS doesn't implement adaptive_avg_pool3d. Add @expectedFailureMPS
decorator to skip the test on MPS devices and fix CI failures.

vishalgoyal316 commented:
@Skylion007, after passing all checks, is this PR ready to merge?

Skylion007 commented on aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:
```diff
@@ -63,19 +65,20 @@ void adaptive_avg_pool3d_out_frame(
                   scalar_t val =
                       *(ip + it * istrideT + ih * istrideH + iw * istrideW);
-                  sum += val;
+                  sum += accscalar_t(val);
```
Actually not sure this cast is necessary if it supports implicit upcasting. Probably fine, but don't want to override this in CUDA as some float type may have specific overloads for reduced precision dtypes.

If this is just to silence warnings/errors probably fine and a super minor nit.

Skylion007 commented on aten/src/ATen/native/AveragePool3d.cpp:
```diff
@@ -396,15 +399,16 @@ void avg_pool3d_backward_out_frame(
             /* scatter gradients out to footprint: */
-            scalar_t val  = *op++;
+            accscalar_t val  = static_cast<accscalar_t>(*op++);
+            accscalar_t val_scaled = val / divide_factor;
```
This value is const and applied to all values in the for loop below, should be statically cast once to scalar_t and applied. Casting in the loop is inefficient.

Skylion007 reviewed (commented):
Finally remaining performance nits

Skylion007 reviewed (commented):
Okay, now I think we have some unnecessary casts and things that can be improved for performance here.

vishalgoyal316 commented:
> Okay, now I think we have some unnecessary casts and things that can be improved for performance here.

@Skylion007, you are right, and thanks for pointing out such improvements, I have pushed the code accordingly, please let me know if anything else need to correct/update.

Skylion007 commented on test/nn/test_pooling.py:
```diff
@@ -1011,6 +1011,118 @@ def test_adaptive_avg_pool3d_output_size_one(self, device):
+            result_reduced.float(), result_fp32, atol=1e-2, rtol=1e-3
+        )
+
+    @onlyNativeDeviceTypes
```
Is this test also failing on MPS?

malfet replied:
+1 on the above, why this decorator is needed here?

vishalgoyal316 replied:
The decorator ensures the test only runs on devices that actually compute values, so it can verify the precision fix works correctly.

malfet commented on test/nn/test_pooling.py:
```diff
@@ -1011,6 +1011,118 @@ def test_adaptive_avg_pool3d_output_size_one(self, device):
+            [
+                [
+                    [
+                        [-1.4062, 1.4609, 0.6797],
```
Once again, what so specific about those values?

vishalgoyal316 replied:
These specific values likely triggered the original bug reported in issue #140696. They represent a real-world case where precision mattered.

malfet commented on aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu:
```diff
@@ -106,7 +106,7 @@ __global__ void adaptiveaveragepool(
                                    (istartT + it) * istrideT +
                                    (istartH + ih) * istrideH + (istartW + iw) * istrideW;
             scalar_t val = input[input_offset];
-            sum += static_cast<accscalar_t>(val);
```
This change is benign, but why is it needed now?

malfet commented on aten/src/ATen/native/AveragePool3d.cpp:
```diff
@@ -396,15 +398,15 @@ void avg_pool3d_backward_out_frame(
                 for (auto x = wstart; x < wend; x++)
                 {
-                  *(ip + z * iheight * iwidth + y * iwidth + x) += val / divide_factor;
+                  *(ip + z * iheight * iwidth + y * iwidth + x) += val_scaled;
```
This is a nice change, but have nothing to do with precision improvements, is it?

malfet reviewed (changes requested):
This PR tries to do many things at once, do you mind breaking it down into smaller changes? (For example small BE changes that precomputes divide factor, another one that extends types, and finally the one that adds new test?

vishalgoyal316 commented:
@pytorchbot rebase
